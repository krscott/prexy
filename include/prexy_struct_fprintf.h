#ifndef PREXY_STRUCT_FPRINTF_H_
#define PREXY_STRUCT_FPRINTF_H_

#include <assert.h>
#include <stdio.h>

// Get printf format string for a type.
// Does *not* assume a `char *` is a `%s`.
#define PREXY_UTIL_SIMPLE_TYPE_FORMAT(x)                                       \
    (_Generic(                                                                 \
        (x),                                                                   \
         _Bool: "%d",                                                          \
         short: "%hd",                                                         \
         unsigned short: "%hu",                                                \
         int: "%d",                                                            \
         unsigned int: "%u",                                                   \
         long: "%ld",                                                          \
         unsigned long: "%lu",                                                 \
         long long: "%lld",                                                    \
         unsigned long long: "%llu",                                           \
         float: "%f",                                                          \
         double: "%f",                                                         \
         char: "'%c'",                                                         \
         void *: "%p"                                                          \
    ))

#define PREXY_STRUCT_FPRINT_REPR_simple(type, varname)                         \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        fprintf(                                                               \
            stream,                                                            \
            PREXY_UTIL_SIMPLE_TYPE_FORMAT((x)->varname),                       \
            (x)->varname                                                       \
        );                                                                     \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_simple_array(type, varname, len)              \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            fprintf(                                                           \
                stream,                                                        \
                PREXY_UTIL_SIMPLE_TYPE_FORMAT((x)->varname[i]),                \
                (x)->varname[i]                                                \
            );                                                                 \
            fprintf(stream, ", ");                                             \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR_struct(type, varname)                         \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        type##_fprint_repr(stream, &(x)->varname);                             \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_struct_array(type, varname, len)              \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            type##_fprint_repr(stream, &(x)->varname[i]);                      \
            fprintf(stream, ", ");                                             \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR_enum(type, varname)                           \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        fprintf(stream, "%s", type##_to_cstr((x)->varname));                   \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_enum_array(type, varname, len)                \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            fprintf(stream, "%s, ", type##_to_cstr((x)->varname[i]));          \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR(fkind, ...)                                   \
    PREXY_STRUCT_FPRINT_REPR_##fkind(__VA_ARGS__)

#define prexy_struct_fprint_repr_decl(name)                                    \
    void name##_fprint_repr(FILE *stream, struct name const *x)
#define prexy_struct_fprint_repr_impl_x(name, FIELDS_X)                        \
    prexy_struct_fprint_repr_decl(name)                                        \
    {                                                                          \
        fprintf(stream, "(struct " #name "){ ");                               \
        FIELDS_X(PREXY_STRUCT_FPRINT_REPR)                                     \
        fprintf(stream, "}");                                                  \
    }                                                                          \
    static_assert(1, "")
#define prexy_struct_fprint_repr_impl(name)                                    \
    prexy_struct_fprint_repr_impl_x(name, name##_x_fields)
#define prexy_struct_fprint_repr_impl_attr(name, attr)                         \
    prexy_struct_fprint_repr_impl_x(name, name##_x_fields_##attr);             \
    static_assert(sizeof(*(struct attr *)0), "")

#endif
