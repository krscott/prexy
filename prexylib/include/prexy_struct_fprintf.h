#ifndef PREXY_STRUCT_FPRINTF_H_
#define PREXY_STRUCT_FPRINTF_H_

#include "prexy.h"
#include <assert.h>
#include <stdio.h>

// Get printf format string for a type.
// Does *not* assume a `char *` is a `%s`.
#define PREXY_UTIL_SIMPLE_TYPE_FORMAT(x)                                       \
    (_Generic(                                                                 \
        (x),                                                                   \
         _Bool: "%d",                                                          \
         short: "%hd",                                                         \
         unsigned short: "%hu",                                                \
         int: "%d",                                                            \
         unsigned int: "%u",                                                   \
         long: "%ld",                                                          \
         unsigned long: "%lu",                                                 \
         long long: "%lld",                                                    \
         unsigned long long: "%llu",                                           \
         float: "%f",                                                          \
         double: "%f",                                                         \
         char: "'%c'",                                                         \
         void *: "%p"                                                          \
    ))

#define PREXY_STRUCT_FPRINT_REPR_simple(type, varname)                         \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        fprintf(                                                               \
            stream,                                                            \
            PREXY_UTIL_SIMPLE_TYPE_FORMAT((x)->varname),                       \
            (x)->varname                                                       \
        );                                                                     \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_simple_array(type, varname, len)              \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            fprintf(                                                           \
                stream,                                                        \
                PREXY_UTIL_SIMPLE_TYPE_FORMAT((x)->varname[i]),                \
                (x)->varname[i]                                                \
            );                                                                 \
            fprintf(stream, ", ");                                             \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR_struct(type, varname)                         \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        type##_fprint_repr(stream, &(x)->varname);                             \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_struct_array(type, varname, len)              \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            type##_fprint_repr(stream, &(x)->varname[i]);                      \
            fprintf(stream, ", ");                                             \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR_enum(type, varname)                           \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        fprintf(stream, "%s", type##_to_cstr((x)->varname));                   \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_enum_array(type, varname, len)                \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            fprintf(stream, "%s, ", type##_to_cstr((x)->varname[i]));          \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR_FIELD_SELECT(fkind, ...)                      \
    PREXY_STRUCT_FPRINT_REPR_##fkind(__VA_ARGS__)

prexy_namespace(prexy_struct_fprint_repr);

#define prexy_struct_fprint_repr_decl(name)                                    \
    void name##_fprint_repr(FILE *stream, struct name const *x)

#define prexy_struct_fprint_repr_impl(name, FIELDS_X)                          \
    prexy_struct_fprint_repr_decl(name)                                        \
    {                                                                          \
        fprintf(stream, "(struct " #name "){ ");                               \
        FIELDS_X(PREXY_STRUCT_FPRINT_REPR_FIELD_SELECT)                        \
        fprintf(stream, "}");                                                  \
    }

#endif
