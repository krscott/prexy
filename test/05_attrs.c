#include "05_attrs_prexy.h"
#include "prexy_struct_fprintf.h"
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <string.h>

#define PREXY_STRUCT_FPRINT_REPR_strattr(type, varname, ...)                   \
    {                                                                          \
        fprintf(stream, "." #varname " = \"%s\", ", (x)->varname);             \
    }

struct strattr
{
    bool is_cstr;
};

prexy struct cstr
{
    px_attr(strattr, .is_cstr = true);
    char const *buf;

    size_t len;
};
static prexy_struct_fprint_repr_impl_attr(cstr, strattr);

prexy struct messages
{
    struct cstr buf[5];
};
static prexy_struct_fprint_repr_impl(messages);

static struct cstr make_cstr(char const *s)
{
    return (struct cstr){
        .buf = s,
        .len = strlen(s),
    };
}

int main(void)
{
    struct cstr s = make_cstr("Hello, World!");
    cstr_fprint_repr(stdout, &s);

    fprintf(stdout, "\n\n");

    struct messages msgs = {
        .buf =
            {
                make_cstr("never"),
                make_cstr("gonna"),
                make_cstr("give"),
                make_cstr("you"),
                make_cstr("up"),
            },
    };
    messages_fprint_repr(stdout, &msgs);
    return 0;
}
