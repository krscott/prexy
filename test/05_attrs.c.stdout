(struct cstr){ .buf = "Hello, World!", .ptr = 0x5618e964d004, .len = 13, }

(struct messages){ .buf = { (struct cstr){ .buf = "never", .ptr = 0x5618e964d015, .len = 5, }, (struct cstr){ .buf = "gonna", .ptr = 0x5618e964d01b, .len = 5, }, (struct cstr){ .buf = "give", .ptr = 0x5618e964d021, .len = 4, }, (struct cstr){ .buf = "you", .ptr = 0x5618e964d026, .len = 3, }, (struct cstr){ .buf = "up", .ptr = 0x5618e964d02a, .len = 2, }, }, }